### ISSUE TITLE
session duration-based asset playback functionality

### ISSUE THREAD

**Issue by hburgund** [roundware/roundware-server#187]
Currently assets are made available for playback based on the user's location and the tag filters provided by the user.  Once made available, there are various ordering and prioritization adjustments that can be made to the ```playlist```.

I want to introduce a new method which enables assets to be available for playback based on the duration of the session and is independent of location.  In other words, no matter where a user walks around, at a given session-duration time, say 2 minutes, a set of assets can be added to whatever other assets are already available per the user's location.  

This allows for some significant new creative and aesthetic opportunities for people setting up RW projects.  Some slightly more time-based narrative elements can be introduced and in situations where a large geographic area is covered, this method would ensure that users experience some asset content even if they don't happen to run into any as they walk around.

Here are my thoughts on how to architect this:

* create a new list in recording_collection called ```playlist_duration```
* add and subtract assets from this list based on some sort of recurring timer to check which assets have a duration-based availability (and the respective start and end times of their availability)
* the queue of assets to be played back is now based on the already existing ```playlist_location``` assets and the new ```playlist_duration``` assets combined (with a duplicates check!).  Perhaps ```playlist_duration``` content is prioritized by default; or maybe a flag to switch between which list is prioritized as I can see scenarios where each prioritization would be useful.
* my current feeling is that the tag filters should be applied to both location and session-duration triggered assets

The database and admin would need to be adjusted as well:

* add one-to-many model on asset model called something like ```asset_playlist_duration``` idk
* fields in this new model are:
  * ```asset_id```
  * ```start_time``` (s)
  * ```end_time``` (s)
* duration-based availability could be an inline on the asset model (like votes) in the admin so that the fields can be added easily for as many time ranges as desired

I haven't really thought this through entirely, but wanted to get my initial ideas recorded as a starting place for this.  I'm sure there are numerous ways to accomplish a similar end result.

**Comment by eosrei** [comment]
Q: If the recording is on the duration playlist will it also be available by location?

**Comment by hburgund** [comment]
I think in most practical situations, an asset that has a geographic location will not also be set to be in the duration playlist.  So if that simplifies things from an implementation standpoint, then we can assume that to be the case always.  But if it's no easier one way or the other, then might as well not restrict in that way.  Happy to take your advice on this from an implementation standpoint, so let me know.

**Comment by eosrei** [comment]
> add and subtract assets from this list based on some sort of recurring timer to check which assets have a duration-based availability (and the respective start and end times of their availability)

It'll be simpler to implement and be less wasteful to update this list only when a new recording is requested.

Here's the model/table I think will cover everything:
```
TimedPlaylist(Model):
    id = Everyone gets unique IDs! You get one and you get one!
    pid = Project ID
    aid = Asset ID
    start = Time in seconds
    end = Time in seconds
```

Pseudocode for RoundwareD.py: 
* Load all TimedPlaylist items for the related Project ID.
* Load all available recordings
* When an asset is requested:
    * Update the list of available Timed Playlist items
        * Get currently timed assets available. End time < Current time and Start Time < Current time.
        * Remove all assets not in the "all assets"(only includes only tag filtered items).
        * Remove all assets in banned_timeout and banned_proximity
        * Sort by project default method?
    * If Timed Playlist has anything in it return the first item (TODO, remove it from the nearby_playlist (yes, a TODO in the pseudocode, It won't get implemented first time through haha))
    * Otherwise, return the first item in the playlist_nearby

I think this cover everything without being super complex...

**Comment by eosrei** [comment]
Blast. I must finish the cleaning the model/schema before I can modify it. Switching to: #165

**Comment by hburgund** [comment]
Generally this looks great, as far as I can tell, though I have a few questions:

* I like the idea of not having a timer but rather only updating the Timed Playlist items when necessary.  I'm not entirely clear on when exactly a new recording is requested, which would trigger this code to run, as you say.  I'm guessing this is triggered on every ```modify_stream```, but since these assets are timed and not related to location, I don't want the whole process to be driven entirely by location changes.  I'm probably just missing something about how new recordings are requested, but this functionality needs to work if the listener isn't moving as well as if the project doesn't even have geo-listen enabled.  Are new recordings requested every time an audiotrack isn't currently playing something and its dead air time runs out?
* it should be: "End time > Current time and Start Time < Current time", yes?
* how will this be represented in the Admin?  Can we do an Inline on the Asset model so that I can add entries in the ```TimedPlaylist``` model right from the Asset in question.  That would be the most intuitive, I think.  Like what is done with the Vote model for Assets.

I have a hard time running through this all in my mind, but I do think your plan sounds like it will achieve the desired functionality, though we may need to tweak during implementation, of course.

**Comment by eosrei** [comment]
>  I'm not entirely clear on when exactly a new recording is requested, which would trigger this code to run, as you say.

When ever a new recording is requested `get_recording()`, we'll update the timed stuff: https://github.com/hburgund/roundware-server/blob/db32a043fa3db7d359b9544083f5f28743de7877/roundwared/recording_collection.py#L76

> Are new recordings requested every time an audiotrack isn't currently playing something and its dead air time runs out?

Yes: https://github.com/hburgund/roundware-server/blob/db32a043fa3db7d359b9544083f5f28743de7877/roundwared/audiotrack.py#L53

> it should be: "End time > Current time and Start Time < Current time", yes?

Ha! Yes. Mistyping...

> how will this be represented in the Admin? 

The initial iteration will not include Admin changes to reduce review/code complexity, but we'll figure a good way to handle it in the UI.

**Comment by hburgund** [comment]
All sounds great for typical scenarios.

That said, how would this, admittedly very edgy, case work: There is a non-geo-listen project with ALL assets in the project assigned as timed assets.  If none of the timed assets are set to begin until say 30 seconds into the stream duration, that means nothing would play until then because nothing is available, but what would then be the trigger for a new recording when 30s rolls around?

For geo-listen projects, this seems like it would never be an issue since lat/lon changes would trigger ```modify_stream``` requests, triggering ```get_recording()```, but for non-geo-listen, it seems like the timed assets might never get triggered in certain setups.  I suppose we could introduce another API call to force-trigger an asset check that would be the client's responsibility to send at whatever frequency made sense for the project, or the server could handle it somehow.

Or maybe I'm over-thinking?  I don't want this edge case to cause any major extra effort, but I like to think about the extremes just to fully understand how something will operate.

**Comment by eosrei** [comment]
> what would then be the trigger for a new recording when 30s rolls around?

The system should keep attempting to start audio tracks. When nothing is available it checks again in a random (within the range) amount of time.

**Comment by eosrei** [comment]
> what would then be the trigger for a new recording when 30s rolls around?
Update, here's the code: https://github.com/roundware/roundware-server/blob/develop/roundwared/audiotrack.py#L69 Every time move_listener is called, if there is no current audio track one is scheduled.

**Comment by eosrei** [comment]
Closed by #217
